### PR TITLE
Add new op: calc_reduced_attention_scores.

### PR DESCRIPTION
**Pull request by umiswing**
### PR Category

Others

### PR Types

New features

### Description

Add a new op: `calc_reduced_attention_scores`, which returns reduce sum of attention scores over dim `seqlen_q`.
related pr: https://github.com/PaddlePaddle/flash-attention/pull/43

### CODE REVIEW

File: python/paddle/nn/functional/flash_attention.py [hunk 1]
```diff
@@ -1251,3 +1251,107 @@ def flash_attention_with_sparse_mask(
...
+        query.stop_gradient and key.stop_gradient
+    ), 'calc_reduced_attention_scores() is for inference only.'
+
+    if in_dynamic_mode():
+        (reduced_scores, _) = _C_ops.calc_reduced_attn_scores(
+            query, key, softmax_lse, False
+        )
+        return reduced_scores
+
+    helper = LayerHelper('calc_reduced_attn_scores', **locals())
```
Comment (SigureMo) [inline python/paddle/nn/functional/flash_attention.py:1336]: 这个 API 可以支持 PIR 么？预计 8 月 develop 就要切 PIR 为默认模式了

File: python/paddle/nn/functional/flash_attention.py [hunk 2]
```diff
@@ -1251,3 +1251,107 @@ def flash_attention_with_sparse_mask(
...
+            >>> dtype = 'float16'
+            >>> query = np.random.random(q_shape)
+            >>> key = np.random.random(k_shape)
+            >>> q = paddle.to_tensor(
+            >>>     query, place=place, dtype=dtype, stop_gradient=True
+            >>> )
+            >>> k = paddle.to_tensor(
+            >>>     key, place=place, dtype=dtype, stop_gradient=True
+            >>> )
+            >>> (_,_,softmax_lse,_) = _C_ops.flash_attn(
```
Comment (SigureMo) [inline python/paddle/nn/functional/flash_attention.py]:
```suggestion
            >>> _, _, softmax_lse, _ = _C_ops.flash_attn(
```

这些地方适当加一些空格

File: python/paddle/nn/functional/flash_attention.py [hunk 3]
```diff
@@ -1251,3 +1251,107 @@ def flash_attention_with_sparse_mask(
...
+            >>>     q,
+            >>>     k,
+            >>>     k,
+            >>>     (None,),#fixed_seed_offset
+            >>>     None,#attn_mask
+            >>>     0.0,#dropout
+            >>>     False,#causal
+            >>>     False,#return_softmax
+            >>>     False,#is_test
+            >>>     ""#rng_name
```
Comment (SigureMo) [inline python/paddle/nn/functional/flash_attention.py]:
```suggestion
            >>>     q,
            >>>     k,
            >>>     k,
            >>>     (None,),  # fixed_seed_offset
            >>>     None,  # attn_mask
            >>>     0.0,  # dropout
            >>>     False,  # causal
            >>>     False,  # return_softmax
            >>>     False,  # is_test
            >>>     "",  # rng_name
```

File: paddle/phi/infermeta/ternary.h [hunk 4]
```diff
@@ -128,6 +128,11 @@ void FlashAttnQKVPackedInferMeta(const MetaTensor& qkv,
                                  MetaTensor* softmax_lse,
                                  MetaTensor* seed_offset);
 
+void CalcReducedAttnScoresInferMeta(const MetaTensor& q,
+                                    const MetaTensor& k,
+                                    MetaTensor* reduced_scores,
+                                    MetaTensor* softmax);
```
Comment (zyfncg) [inline paddle/phi/infermeta/ternary.h]: 两个输入tensor应该属于binary分类

File: paddle/phi/infermeta/ternary.cc [hunk 5]
```diff
@@ -515,6 +515,34 @@ void FlashAttnQKVPackedInferMeta(const MetaTensor& qkv,
...
+  int seqlen_q = q.dims()[1];
+  int seqlen_k = k.dims()[1];
+  if (reduced_scores) {
+    reduced_scores->set_dtype(phi::DataType::FLOAT32);
+    reduced_scores->set_dims({batch_size, num_heads, 1, seqlen_k});
+  }
+  if (softmax) {
+    softmax->set_dtype(phi::DataType::FLOAT32);
+    softmax->set_dims({batch_size, num_heads, seqlen_q, seqlen_k});
+  }
```
Comment (zyfncg) [inline paddle/phi/infermeta/ternary.cc]: 这两个输出结果是optional的吗？

File: python/paddle/nn/functional/flash_attention.py [hunk 6]
```diff
@@ -1251,3 +1251,107 @@ def flash_attention_with_sparse_mask(
...
+    reduced_scores = helper.create_variable_for_type_inference(paddle.float32)
+    softmax = helper.create_variable_for_type_inference(paddle.float32)
+    inputs = {
+        'q': query,
+        'k': key,
+        'softmax_lse': softmax_lse,
+    }
+    outputs = {
+        'reduced_scores': reduced_scores,
+        'softmax': softmax,
```
Comment (zyfncg) [inline python/paddle/nn/functional/flash_attention.py]: softmax没有作为最终返回结果，是否有必要作为输出？

File: paddle/phi/infermeta/ternary.cc (hunk 5 above)
Comment (umiswing) [reply paddle/phi/infermeta/ternary.cc]: > 这两个输出结果是optional的吗？

@zyfncg 
`reduced_scores`不是optional的。
`softmax`我想还是从接口中去除更好。

File: python/paddle/nn/functional/flash_attention.py (hunk 6 above)
Comment (umiswing) [reply python/paddle/nn/functional/flash_attention.py]: > softmax没有作为最终返回结果，是否有必要作为输出？

@zyfncg 
好的，确实是从接口中去除`softmax`更好。

File: paddle/phi/infermeta/ternary.h (hunk 4 above)
Comment (umiswing) [reply paddle/phi/infermeta/ternary.h]: > 两个输入tensor应该属于binary分类

@zyfncg 
实际上是三个输入tensor: `q`, `k`, `softmax_lse`。`q` `k`足够推导出输出tensor的形状。不过，我在这个InferMeta里添加对输入tensor的检查吧。

File: python/paddle/nn/functional/flash_attention.py (hunk 1 above)
Comment (umiswing) [reply python/paddle/nn/functional/flash_attention.py:1336]: > 这个 API 可以支持 PIR 么？预计 8 月 develop 就要切 PIR 为默认模式了

@SigureMo 我们暂时没有在PIR模式下使用这个API的场景。若需添加PIR的支持，可否在下个PR中单独添加？

Comment (SigureMo) [reply python/paddle/nn/functional/flash_attention.py:1336]: 这里 `in_dynamic_mode` 改成 `in_dynamic_or_pir_mode` 应该就可以了？看样子也是动静统一的 API，然后开启 `FLAGS_enable_pir_api=True` 验证一下没问题就行了

如果有问题，那下个 PR 再单独适配吧

**Review by SigureMo — DISMISSED** [review]
LGTMeow  for new API type annotations

麻烦 @sunzhongkai588 review 下新增 API 的文档

File: python/paddle/nn/functional/flash_attention.py (hunk 1 above)
Comment (umiswing) [reply python/paddle/nn/functional/flash_attention.py:1336]: > 这里 `in_dynamic_mode` 改成 `in_dynamic_or_pir_mode` 应该就可以了

done

File: python/paddle/nn/functional/flash_attention.py [hunk 7]
```diff
@@ -1251,3 +1251,103 @@ def flash_attention_with_sparse_mask(
...
+                        [batch_size, seqlen_k, num_heads, head_dim].
+                        The dtype can be float16 or bfloat16.
+        softmax_lse(Tensor): The logsumexp of each row returned by _C_ops.flash_attn().
+                        3-D tensor with shape:
+                        [batch_size, num_heads, seqlen_q].
+                        The dtype is float32.
+    Returns:
+        reduced_attention_scores(Tensor), The reduce sum of attention scores across seqlen_q.
+                    4-D tensor with shape: [batch_size, num_heads, 1, seqlen_k].
+                    The dtype is float32.
```
Comment (sunzhongkai588) [inline python/paddle/nn/functional/flash_attention.py]:
```suggestion
    Returns:
        reduced_attention_scores(Tensor), The reduce sum of attention scores across seqlen_q.
        4-D tensor with shape ``[batch_size, num_heads, 1, seqlen_k]``. The dtype is float32.
```
官网渲染出了点问题
image

**Review by zyfncg — DISMISSED** [review]

**Review by sunzhongkai588 — DISMISSED** [review]
LGTM

**Review by wanghuancoder — DISMISSED** [review]
LGTM

**Review by SigureMo — DISMISSED** [review]

File: python/paddle/nn/functional/__init__.py [hunk 8]
```diff
@@ -291,5 +292,6 @@
     'flash_attention_with_sparse_mask',
     'flash_attn_qkvpacked',
     'flash_attn_varlen_qkvpacked',
+    'calc_reduced_attention_scores',
```
Comment (jeff41404) [inline python/paddle/nn/functional/__init__.py]: is `paddle.nn.functional.calc_reduced_attention_scores` public API? If yes, according to the newly added API process, a design document and an Chinese document of official website are required in description; If not, it is only for internal use within the framework, then remove it from the list of __ all__

Comment (umiswing) [reply python/paddle/nn/functional/__init__.py]: > is `paddle.nn.functional.calc_reduced_attention_scores` public API? If yes, according to the newly added API process, a design document and an Chinese document of official website are required in description; If not, it is only for internal use within the framework, then remove it from the list of __ all__

we decide not to make `paddle.nn.functional.calc_reduced_attention_scores` public.

**Review by wanghuancoder — APPROVED** [review]
LGTM

**Review by zyfncg — APPROVED** [review]

**Review by XieYunshen — APPROVED** [review]
LGTM

**Review by GuoxiaWang — APPROVED** [review]
LGTM
